hedera-node/src/main/java/com/hedera/services/fees/calculation/contract/queries/ContractCallLocalResourceUsage.java

```diff
@@ -0,0 +1,121 @@
+import com.hederahashgraph.api.proto.java.ResponseHeader;
+import com.hederahashgraph.api.proto.java.ResponseType;
+import com.hederahashgraph.fee.SmartContractFeeBuilder;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.time.Instant;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.hedera.services.queries.AnswerService.NO_QUERY_CTX;
+import static com.hedera.services.queries.contract.ContractCallLocalAnswer.CONTRACT_CALL_LOCAL_CTX_KEY;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+import static com.hederahashgraph.api.proto.java.ResponseType.ANSWER_ONLY;
+
+public class ContractCallLocalResourceUsage implements QueryResourceUsageEstimator {
+	private static final Logger log = LogManager.getLogger(ContractCallLocalResourceUsage.class);
+
+	private final ContractCallLocalAnswer.LegacyLocalCaller delegate;
+	private final SmartContractFeeBuilder usageEstimator;
+	private final GlobalDynamicProperties properties;
+
+	public ContractCallLocalResourceUsage(
+			ContractCallLocalAnswer.LegacyLocalCaller delegate,
+			SmartContractFeeBuilder usageEstimator,
+			GlobalDynamicProperties properties
+	) {
+		this.delegate = delegate;
+		this.properties = properties;
+		this.usageEstimator = usageEstimator;
+	}
+
+	@Override
+	public boolean applicableTo(Query query) {
+		return query.hasContractCallLocal();
+	}
+
+	@Override
+	public FeeData usageGiven(Query query, StateView view) {
+		return usageFor(query, view, query.getContractCallLocal().getHeader().getResponseType(), NO_QUERY_CTX);
+	}
+
+	@Override
+	public FeeData usageGivenType(Query query, StateView view, ResponseType type) {
+		return usageFor(query, view, type, NO_QUERY_CTX);
+	}
+
+	@Override
+	public FeeData usageGiven(Query query, StateView view, Map<String, Object> queryCtx) {
+		return usageFor(
+				query,
+				view,
+				query.getContractCallLocal().getHeader().getResponseType(),
+				Optional.of(queryCtx));
+	}
+
+	private FeeData usageFor(Query query, StateView view, ResponseType type, Optional<Map<String, Object>> queryCtx) {
+		try {
+			var op = query.getContractCallLocal();
+			ContractCallLocalResponse response;
+			if (queryCtx.isEmpty()) {
+				response = dummyResponse(op.getContractID());
+			} else {
+				response = delegate.perform(op, Instant.now().getEpochSecond());
+				queryCtx.get().put(CONTRACT_CALL_LOCAL_CTX_KEY, response);
+			}
+			var nonGasUsage = usageEstimator.getContractCallLocalFeeMatrices(
+					op.getFunctionParameters().size(),
+					response.getFunctionResult(),
+					type);
+			var ans = nonGasUsage.toBuilder()
+					.setNodedata(nonGasUsage.getNodedata().toBuilder().setGas(op.getGas()))
+					.build();
+			return ans;
+		} catch (Exception e) {
+			log.warn("Usage estimation unexpectedly failed for {}!", query, e);
+			throw new IllegalArgumentException(e);
+		}
+	}
+
+	ContractCallLocalResponse dummyResponse(ContractID target) {
```

Comment:
This method is only used in this class, should it  be `private` ?

Reply:
I left the access level `package-private` because I wanted to call this method directly in tests.

Reply:
I noticed the unit test case you are calling this methods directly and I thought this method might be general util method used by other classes in this package as well.